File: src/main/java/at/jku/ssw/coco/intellij/psi/impl/CocoReferencingType.java
package at.jku.ssw.coco.intellij.psi.impl;

import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.lang.ASTNode;
import com.intellij.psi.ContributedReferenceHost;
import com.intellij.psi.PsiReference;
import com.intellij.psi.PsiReferenceService;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Created by Thomas on 27/03/2015.
 */
public abstract class CocoReferencingType extends ASTWrapperPsiElement implements ContributedReferenceHost {

    public CocoReferencingType(@NotNull ASTNode node) {
        super(node);
    }

    @Nullable
    @Override
    public PsiReference getReference() {
        PsiReference[] references = getReferences();

        if (references.length == 0) {
            return null;
        }

        return references[0];
    }

    @NotNull
    @Override
    public PsiReference[] getReferences() {
        return PsiReferenceService.getService().getContributedReferences(this);
    }
}

File: src/main/kotlin/at/scheinecker/intellij/coco/psi/impl/CocoReferencingType.kt
package at.scheinecker.intellij.coco.psi.impl

import com.intellij.extapi.psi.ASTWrapperPsiElement
import com.intellij.lang.ASTNode
import com.intellij.psi.ContributedReferenceHost
import com.intellij.psi.PsiNameIdentifierOwner
import com.intellij.psi.PsiReference
import com.intellij.psi.PsiReferenceService

/**
 * Created by Thomas on 27/03/2015.
 */
abstract class CocoReferencingType(node: ASTNode) : ASTWrapperPsiElement(node), ContributedReferenceHost, PsiNameIdentifierOwner {

    override fun getReference(): PsiReference? {
        return references.firstOrNull()
    }

    override fun getReferences(): Array<PsiReference> {
        return PsiReferenceService.getService().getContributedReferences(this)
    }
}
